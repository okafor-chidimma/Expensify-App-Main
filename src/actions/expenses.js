import uuid from 'uuid';
// Add expense
export const addExpense = ({ amount = 0, createdAt = 0, note = '', description = '' } = {}) => (
  {
    type: 'ADD_EXPENSE',
    expense: {
      id: uuid(),
      amount,
      createdAt,
      description,
      note
    }
  }
);

// remove expense

export const removeExpense = ({ id }) => {
  return {
    type: 'REMOVE_EXPENSE',
    id
  }
};

// Edit Expense

export const editExpense = ({ id }, updatedExpense) => {
  return {
    type: 'EDIT_EXPENSE',
    id,
    updatedExpense
  }
};
