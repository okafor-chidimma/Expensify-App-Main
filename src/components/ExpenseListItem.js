import React from 'react';
import { Link } from 'react-router-dom';

const ExpenseListItem = ({ description, amount, createdAt, id }) => (
  <div>

    <Link to={`/edit/${id}`}>
      <h4>{description}</h4>
    </Link>
    <p>Amount Spent: {amount}</p>
    <p>Created At: {createdAt}</p>
  </div>
);

export default ExpenseListItem;