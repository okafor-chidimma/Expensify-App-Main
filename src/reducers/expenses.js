// Expenses Reducer
const expenseStateDefault = [];
const expenseReducer = (state = expenseStateDefault, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [
        ...state,
        action.expense
      ];
    case 'REMOVE_EXPENSE':
      return state.filter(({ id }) => id !== action.id);
    case 'EDIT_EXPENSE':
      return state.map((expense) => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.updatedExpense
          }
        } else {
          return expense
        }
      });
    default:
      return state;
  }
};

export default expenseReducer;