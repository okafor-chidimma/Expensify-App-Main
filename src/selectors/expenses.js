import moment from 'moment';

// to sort the expense and return data that matches
const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {

  return expenses.filter((expense) => {
    const createdAtMoment = moment(expense.createdAt);
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment) : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment) : true;
    // this includes will always be true as long as the text is an empty string or a string contained in description
    // i.e textMatch will only be false is the text is not empty and cannot be found in description
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
    // console.log(textMatch, 'text match');
    // console.log(endDate, 'end date');
    // console.log(createdAtMoment, 'created at moment');
    // console.log(expense.createdAt, 'value of');
    return startDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1
    } else if (sortBy === 'amount') {
      return a.amount < b.amount ? 1 : -1
    }
  });

}

export default getVisibleExpenses;