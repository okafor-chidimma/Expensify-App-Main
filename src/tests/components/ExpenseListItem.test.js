import React from "react";
import { shallow } from "enzyme";
import ExpenseListItem from "../../components/ExpenseListItem";
import expenses from "../fixtures/expenses";

const [expense] = expenses;
describe("Rendering <ExpenseListItem/>", () => {
  it("Render the <ExpenseListItem/> with expense as props", () => {
    const wrapper = shallow(<ExpenseListItem {...expense} />);
    expect(wrapper).toMatchSnapshot();
  });
});
