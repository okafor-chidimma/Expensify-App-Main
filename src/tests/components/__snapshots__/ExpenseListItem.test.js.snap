// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering <ExpenseListItem/> Render the <ExpenseListItem/> with expense as props 1`] = `
<div>
  <Link
    to="/edit/1"
  >
    <h4>
      January rent
    </h4>
  </Link>
  <p>
    Amount Spent: 
    54500
  </p>
  <p>
    Created At: 
    0
  </p>
</div>
`;
